### DISCUSSION TITLE
Suggest a strong Content Security Policy (CSP) rule by default

### DISCUSSION THREAD

**Discussion by toastal** [srid/emanote#375] [Ideas]
Currently the CSP for emanote is approximately (visiting one page with [Laboratory](https://addons.mozilla.org/en-US/firefox/addon/laboratory-by-mozilla/)) 

```
default-src 'none'; font-src 'self'; img-src 'self' data:; script-src 'self' 'unsafe-inline' https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.6.0/highlight.min.js https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.6.0/languages/haskell.min.js https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.6.0/languages/nix.min.js; style-src 'self' 'unsafe-inline' https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.6.0/styles/ 
```

I would like to see this as recommended by most security folks.

```
default-src 'none'; font-src 'self'; img-src 'self' https:; script-src 'self'; style-src 'self'
```

But due to a Firefox bug with SVG sprites (`<symbol>` + `<use>`), this is an acceptable alternative for now

```
default-src 'self'; img-src 'self' https:; object-src 'none'; frame-ancestors 'none'
```

---

Motivation: a lot of folks forget this big security win for their content and by recommending a good, strong list, you can steer users into making good security decisions. You can also help folks get better security scores at places like [Mozilla Observatory](https://observatory.mozilla.org/).

---

This change would affect this project too—ultimately in a positive direction in my opinion, especially when we look at how less-than-stellar Emanote’s score is: 
**F** (https://observatory.mozilla.org/analyze/emanote.srid.ca).

By removing the CDN usage, users will not need to leak data to third-parties. Currenty said scripts aren’t even using [subresource integrity](https://developer.mozilla.org/en-US/docs/Web/Security/Subresource_Integrity) to help mitigate a corrupt/compromised CDN pushing out bad content. Since the only usage is for highlight.js, this can be a catalyst to do all syntax highlighting server side. Server side rendering would remove any need for any [CDN trust](https://httptoolkit.com/blog/public-cdn-risks/), but it’s also a massive performance win for downstream users—all of which are downloading these scripts and parsing the same source code lines on their device wasting bandwidth, energy, render speed—when the build process can do it once for everyone. If something is used for LaTex/MathJax, same logic for server-side rendering should apply. As a worst case, at least with `self`, the CSP would encourage folks into self-hosting their libraries for the benefit of their reader.

Unsafe-inline is removed too, which should provoke the project to bundle up its scripts/styles/images into separate files for caching.

**Comment by srid** [comment]
These are generally worth doing.

The https redirect thing is broken for me since eternity; from https://github.com/EmaApps/emanote/settings/pages:
<img width="732" alt="image" src="https://user-images.githubusercontent.com/3998/201786682-d9aaef67-fe64-47d5-a4d4-0f10ef07393b.png">

**Reply by toastal** [reply]
Whaaa. I've never seen that error.
